### PR TITLE
[1.18] Add MobFinalizeSpawnEvent

### PR DESCRIPTION
**Pull request by seymourimadeit**
This PR adds the MobFinalizeSpawnEvent, which allows things to be added to an entity after it has initially spawned. (ie, not loaded from the disk).

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/entity/Mob.java.patch [hunk 1]
```diff
@@ -32,3 +32,13 @@
           if (entity != null) {
              double d0 = entity.m_20280_(this);
              int i = this.m_6095_().m_20674_().m_21611_();
+@@ -988,7 +_,8 @@
+       } else {
+          this.m_21559_(false);
+       }
+-
++      net.minecraftforge.event.entity.EntityFinalizeSpawnEvent e = new net.minecraftforge.event.entity.EntityFinalizeSpawnEvent(this, p_21435_, p_21436_, p_21437_, p_21438_);
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e);
```
Comment (SmellyModder) [inline patches/minecraft/net/minecraft/world/entity/Mob.java.patch]: This can just be one line.
`net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityFinalizeSpawnEvent(this, p_21435_, p_21436_, p_21437_, p_21438_));`

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import net.minecraft.world.entity.SpawnGroupData;
+
+/**
+ * This event is called when an entity finalizes its spawn, usually to give it something (eg, skeletons with bows)<br>
+ * This event should be used if a modder wants to add something to an entity the first time it's spawned.<br>
+ * If a modder wants to add something everytime an entity is spawned they should use {@link EntityJoinWorldEvent}
+ */
+public class EntityFinalizeSpawnEvent extends EntityEvent
+{
+
```
Comment (SmellyModder) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: Remove this blank line.
```suggestion
```

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+{
+    
+    private final DifficultyInstance difficultyInstance;
+    private final MobSpawnType mobSpawnType;
+    @Nullable
+    private final SpawnGroupData groupData;
+    @Nullable
+    private final CompoundTag compoundTag;
+
+    public EntityFinalizeSpawnEvent(Mob entity, DifficultyInstance difficultyInstance, MobSpawnType mobSpawnType, SpawnGroupData data, CompoundTag compoundTag)
```
Comment (SmellyModder) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: `@Nullable` should also be included in the nullable parameters.
```suggestion
    public EntityFinalizeSpawnEvent(Mob entity, DifficultyInstance difficultyInstance, MobSpawnType mobSpawnType, @Nullable SpawnGroupData data, @Nullable CompoundTag compoundTag)
```

**Comment by Shadows-of-Fire** [comment]
What is the desired use case of this event?

**Comment by seymourimadeit** [comment]
> What is the desired use case of this event?

The desired use case is for modders to use this event is if they want to modify an entity the first time they spawn. A modder might not want to use the EntityJoinWorldEvent because that event fires every time a mob joins a world, or use LivingSpawnEvent's subevents because those only fire when an entity is naturally spawned.

File: patches/minecraft/net/minecraft/world/entity/Mob.java.patch [hunk 4]
```diff
@@ -32,3 +32,12 @@
           if (entity != null) {
              double d0 = entity.m_20280_(this);
              int i = this.m_6095_().m_20674_().m_21611_();
+@@ -988,7 +_,7 @@
+       } else {
+          this.m_21559_(false);
+       }
+-
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityFinalizeSpawnEvent(this, p_21435_, p_21436_, p_21437_, p_21438_));
```
Comment (sciwhiz12) [inline patches/minecraft/net/minecraft/world/entity/Mob.java.patch]: While a nitpick, you can shorten this patch by inserting the line and leaving the blank line unaffected.

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package net.minecraftforge.event.entity;
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: Missing license header, please run `updateLicenses`/`licenseFormat`.

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    @Nullable
+    public SpawnGroupData getGroupData() 
+    {
+        return groupData;
+    }
+
+    @Nullable
+    public final CompoundTag getCompoundTag()
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: Why is this getter `final`?

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+public class EntityFinalizeSpawnEvent extends EntityEvent
+{
+    private final DifficultyInstance difficultyInstance;
+    private final MobSpawnType mobSpawnType;
+    @Nullable
+    private final SpawnGroupData groupData;
+    @Nullable
+    private final CompoundTag compoundTag;
+
+    public EntityFinalizeSpawnEvent(Mob entity, DifficultyInstance difficultyInstance, MobSpawnType mobSpawnType, @Nullable SpawnGroupData data, @Nullable CompoundTag compoundTag)
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: In the interests of convenience of avoiding a cast/pattern-matching `instanceof`, please override the `getEntity` getter and change the return type to `Mob`.[^1]

Also, is this event intended for _any_ `Entity` while finalizing spawning, or only for `Mob`s? If only for `Mob`s, it may be beneficial to rename this event to `MobFinalizeSpawnEvent`, to clarify that this only fires for mobs. If not, then the type of the first parameter should be expanded to `Entity` to signal that it may not necessarily be a `Mob` stored in this event.

[^1]: This is possible because [return types are _covariant_](https://www.baeldung.com/java-covariant-return-type) in Java 5+ -- that is, an override can narrow down its return type to be a subclass of the overriden method's return type.

File: src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.DifficultyInstance;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.MobSpawnType;
+import net.minecraft.world.entity.SpawnGroupData;
+
+/**
+ * This event is called when an entity finalizes its spawn, usually to give it something (eg, skeletons with bows)<br>
+ * This event should be used if a modder wants to add something to an entity the first time it's spawned.<br>
+ * If a modder wants to add something everytime an entity is spawned they should use {@link EntityJoinWorldEvent}
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/EntityFinalizeSpawnEvent.java]: Rather then `everytime an entity is spawned`, it may be more understandable if it were reworded to `everytime an entity is added to the world`, as "spawned" is more likely to be understood in the context of "being spawned by something, like a spawner, spawn egg, etc."

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 9]
```diff
@@ -84,7 +84,6 @@
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.Level;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.world.level.biome.MobSpawnSettings;
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Please do not modify unrelated files.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/event/entity/MobFinalizeSpawnEvent.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    @Nullable
+    public CompoundTag getCompoundTag()
+    {
+        return compoundTag;
+    }
+    
+    @Override
+    public Mob getEntity()
+    {
+        return mob;
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/MobFinalizeSpawnEvent.java]: You can avoid storing the `Mob` instance in a field by replacing this with `(Mob) super.getEntity()` instead. IIRC, the compiler is smart enough to know that it won't cause an unchecked cast and therefore not emit a warning for such.

File: src/main/java/net/minecraftforge/event/entity/MobFinalizeSpawnEvent.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    @Nullable
+    public CompoundTag getCompoundTag()
+    {
+        return compoundTag;
+    }
+    
+    @Override
+    public Mob getEntity()
+    {
+        return (Mob)super.getEntity();
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/event/entity/MobFinalizeSpawnEvent.java]: A very small formatting nitpick: space between cast and expression.
```suggestion
        return (Mob) super.getEntity();
```

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
A small formatting nitpick, and then this looks fine to me for assignment. (Please re-request my review once you've applied the change.) Also, please rename your PR's title to match the event's new name.

**Review by sciwhiz12 — APPROVED** [review]

**Comment by TheCurle** [comment]
The idea is great, but the API itself is undocumented and we haven't heard from the author in a while, so I'm going to close this.

If somebody wants to make something similar to allow modifying mobs on first spawn, and document it properly, then feel free.
